aptos-move/aptos-vm/src/move_vm_ext/aggregator_extension.rs

```diff
@@ -0,0 +1,722 @@
+    let aggregator_ref = pop_arg!(args, StructRef);
+    let (handle, key, limit) = get_aggregator_fields(&aggregator_ref)?;
+    let id = AggregatorID::new(handle, key);
+
+    // Get aggregator.
+    let aggregator_context = context.extensions().get::<NativeAggregatorContext>();
+    let mut aggregator_data = aggregator_context.aggregator_data.borrow_mut();
+    let aggregator = aggregator_data.get_aggregator(id, limit);
+
+    aggregator.add(value)?;
+
+    // TODO: charge gas properly.
+    Ok(NativeResult::ok(0, smallvec![]))
+}
+
+/// Move signature:
+/// fun read(aggregator: &Aggregator): u128;
+fn native_read(
+    context: &mut NativeContext,
+    _ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
+    if !cfg!(any(test, feature = "aggregator-extension")) {
+        return Err(not_supported_error());
+    }
+    assert!(args.len() == 1);
+    let aggregator_ref = pop_arg!(args, StructRef);
+
+    // Extract fields from aggregator struct reference.
+    let (handle, key, limit) = get_aggregator_fields(&aggregator_ref)?;
+    let id = AggregatorID::new(handle, key);
+
+    // Get aggregator.
+    let aggregator_context = context.extensions().get::<NativeAggregatorContext>();
+    let mut aggregator_data = aggregator_context.aggregator_data.borrow_mut();
+    let aggregator = aggregator_data.get_aggregator(id, limit);
+
+    // Materialize the value.
+    aggregator.materialize(aggregator_context, &id)?;
+
+    // TODO: charge gas properly.
+    Ok(NativeResult::ok(
+        0,
+        smallvec![Value::u128(aggregator.value)],
+    ))
+}
+
+/// Move signature:
+/// fun sub(aggregator: &mut Aggregator, value: u128);
+fn native_sub(
+    context: &mut NativeContext,
+    _ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
+    if !cfg!(any(test, feature = "aggregator-extension")) {
+        return Err(not_supported_error());
+    }
+    assert!(args.len() == 2);
+
+    // Get aggregator fields and a value to subtract.
+    let value = pop_arg!(args, u128);
+    let aggregator_ref = pop_arg!(args, StructRef);
+    let (handle, key, limit) = get_aggregator_fields(&aggregator_ref)?;
+    let id = AggregatorID::new(handle, key);
+
+    // Get aggregator.
+    let aggregator_context = context.extensions().get::<NativeAggregatorContext>();
+    let mut aggregator_data = aggregator_context.aggregator_data.borrow_mut();
+    let aggregator = aggregator_data.get_aggregator(id, limit);
+
+    // For first version of `Aggregator` (V1), subtraction always materializes
+    // the value first. While this limits commutativity, it is sufficient for
+    // now.
+    // TODO: change this when we implement commutative subtraction.
+    aggregator.materialize(aggregator_context, &id)?;
+    aggregator.sub(value)?;
+
+    // TODO: charge gas properly.
+    Ok(NativeResult::ok(0, smallvec![]))
+}
+
+/// Move signature:
+/// fun remove_aggregator(table_handle: u128, key: u128);
```

Comment:
why this doesn't take an Aggregator instead?

Reply:
No specific reason - changed that:)